### ISSUE TITLE
"local" doesn't work in theme definitions

### ISSUE THREAD

**Issue by apjanke** [robbyrussell/oh-my-zsh#3742]
A lot of the theme definition files (`themes/*.zsh-theme`) are using the `local` declaration to make their variables local to the theme definition. This doesn't actually work, because `local` variables in `zsh` are dynamically scoped, not lexically scoped, and the theme definition files are run using `source` so they're evaluated in the caller's function stack frame. (A sourced file is not a scope, unlike autoloaded functions.) In the current code's design, that's the global namespace. So those `local` declarations are having no effect.

This could be fixed by doing the `source` for theme loading inside a function call to provide a level of indirection for the dynamic scoping. Could be done with an anonymous function or a specially-provided function (which could be more readable for debugging).

Same thing probably applies to plugins and `lib/*.zsh` files if they're using `local` at the top level. You could wrap most of `oh-my-zsh.sh` in a function call to provide a scope frame if you wanted to take care of it all in one fell swoop.

### Affected themes

Here's a quick list of themes that look like they're using `local`. It's not fully correct, because they could be using it inside functions. But I spot-checked several of them and they just had bare `local` declarations at the top level.

```
$ grep -rl local themes  | wc -l
      51
$ grep -rl local themes
themes/af-magic.zsh-theme
themes/afowler.zsh-theme
themes/agnoster.zsh-theme
themes/alanpeabody.zsh-theme
themes/avit.zsh-theme
themes/bira.zsh-theme
themes/bureau.zsh-theme
themes/candy-kingdom.zsh-theme
themes/dieter.zsh-theme
themes/dogenpunk.zsh-theme
themes/dpoggi.zsh-theme
themes/eastwood.zsh-theme
themes/essembeh.zsh-theme
themes/fino-time.zsh-theme
themes/fino.zsh-theme
themes/fishy.zsh-theme
themes/flazz.zsh-theme
themes/frontcube.zsh-theme
themes/funky.zsh-theme
themes/gallifrey.zsh-theme
themes/gallois.zsh-theme
themes/gnzh.zsh-theme
themes/intheloop.zsh-theme
themes/itchy.zsh-theme
themes/jnrowe.zsh-theme
themes/jonathan.zsh-theme
themes/jreese.zsh-theme
themes/kennethreitz.zsh-theme
themes/linuxonly.zsh-theme
themes/lukerandall.zsh-theme
themes/macovsky-ruby.zsh-theme
themes/macovsky.zsh-theme
themes/mira.zsh-theme
themes/mortalscumbag.zsh-theme
themes/murilasso.zsh-theme
themes/muse.zsh-theme
themes/obraun.zsh-theme
themes/peepcode.zsh-theme
themes/pure.zsh-theme
themes/pygmalion.zsh-theme
themes/re5et.zsh-theme
themes/robbyrussell.zsh-theme
themes/simonoff.zsh-theme
themes/smt.zsh-theme
themes/sonicradish.zsh-theme
themes/sorin.zsh-theme
themes/sunaku.zsh-theme
themes/sunrise.zsh-theme
themes/trapd00r.zsh-theme
themes/wedisagree.zsh-theme
themes/ys.zsh-theme
```

**Comment by ncanceill** [comment]
@apjanke you are totally correct. When sourcing a theme (any script, really), only local variables enclosed within functions will really be kept local (as in "local _to the function_"), whereas those declared directly in the script will be kept local _to the shell_ from which you sourced the script.

As you pointed out, some of the themes (_eg_ [agnoster](https://github.com/robbyrussell/oh-my-zsh/blob/master/themes/agnoster.zsh-theme)) do it the right way, and only use `local` within functions.

For the others, it plays out as you would expect. For example, when using [alanpeabody](https://github.com/robbyrussell/oh-my-zsh/blob/master/themes/alanpeabody.zsh-theme), which defines a local `$user` variable directly in the script's scope, you get:
```zsh
echo $user
%{$fg[magenta]%}%n@%{$fg[magenta]%}%m%{$reset_color%}
```

The easy fix is to wrap all offending declarations (and all subsequent code using them) in [anonymous functions](http://zsh.sourceforge.net/Doc/Release/Functions.html#Anonymous-Functions). For example, with the above problematic theme we would have:
```diff
diff --git a/themes/alanpeabody.zsh-theme b/themes/alanpeabody.zsh-theme
index 4a1b1b3..2e0f49d 100644
--- a/themes/alanpeabody.zsh-theme
+++ b/themes/alanpeabody.zsh-theme
@@ -1,4 +1,4 @@
-
+function () {
 local user='%{$fg[magenta]%}%n@%{$fg[magenta]%}%m%{$reset_color%}'
 local pwd='%{$fg[blue]%}%~%{$reset_color%}'
 local rvm=''
@@ -26,3 +26,4 @@ ZSH_THEME_GIT_PROMPT_UNTRACKED="%{$fg[cyan]%} ✭"
 
 PROMPT="${user} ${pwd}$ "
 RPROMPT="${return_code} ${git_branch} ${rvm}"
+}
```

***

There is also the issue of themes not bothering with `local` at all, and just defining global environment variables.

Maybe @robbyrussell would consider updating the contributing guidelines to include something about not leaking (too many) variables to the user's shell?

**Comment by apjanke** [comment]
One minor addition: for scripts that declare `local` directly in the script instead of inside the functions they define, the variables are local to the shell *or function activation* that they are sourced from. If the `source` call happens inside a function, that function call is what they're local to, and their old values will be restored when that function returns.

I took a stab at this in #3743: instead of doing anonymous functions inside the individual themes, I wrapped the `source` call inside a function. Similar effect; it keeps the `local` variables local to the function activation and out of the caller's space.

But just wrapping them in a function wasn't enough: many of them are written to rely on the variables being non-local, and available each time a prompt is issued. The `alanpeabody` theme works fine because it uses double-quotes for the prompt. But another common pattern is this:

```
local ret_status="%(?:%{$fg_bold[green]%}➜ :%{$fg_bold[red]%}➜ %s)"
PROMPT='${ret_status}%{$fg_bold[green]%}%p %{$fg[cyan]%}%c ...more stuff...'
```

The single quotes around the value of `PROMPT` mean that `${ret_status}` isn't actually evaluated until prompt expansion happens when displaying each prompt. And if that `local` had scoped `ret_status` to just the prompt initialization `source` call, then `$ret_status` is undefined at prompt expansion time, so the green/red arrow at the beginning of the prompt doesn't get displayed.

This case seems like an easy fix; since `$ret_status` doesn't depend on any variables which change over time, it can just be pulled out of the quotes for eager evaluation.

```
PROMPT=${ret_status}'%{$fg_bold[green]%}%p %{$fg[cyan]%}%c ...more stuff...'
```

Other themes require a bit more modification of their prompt strings to get them to work right. I tweaked the themes that did this in #3743 to work within a scope that makes `local` local. I also added a `$ZSH_THEME_DEBUG` setting which lets you see what parameters were changed or leaked by the theme.

I'm not sure which way is better. Changing individual themes to use anonymous functions would be the smallest change from the current state. But wrapping the `source` in a function call would give you the convenience of being able to use `local` at the top level of the theme definitions. And it would give you the flexibility to be able to load themes from within functions, where the current design requires that themes be sourced from outside a function, because it's easy to write themes that accidentally rely on that behavior.
